libc/string_mem.c

```diff
@@ -0,0 +1,391 @@
+	TEST_ASSERT_EQUAL_PTR(dst, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, dst, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, src, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, src + (TEST_STR_SIZE * 2), MEM_SIZE - (OFFSET + (TEST_STR_SIZE * 2)));
+}
+
+
+TEST(string_memmove, whole_overlap)
+{
+	uint8_t *const src = mem + OFFSET;
+	/* Move at the same address as block is */
+	ret = memmove(src, src, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(src, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, src, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, src + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, right_overlap)
+{
+	uint8_t *const src = mem + OFFSET;
+	uint8_t *dst = mem + (OFFSET + OVERLAP_LEN);
+	/* Move data so that data blocks overlap */
+	ret = memmove(dst, src, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(dst, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, dst, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, src, TEST_STR_SIZE - OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, src + TEST_STR_SIZE + OVERLAP_LEN, MEM_SIZE - (OFFSET + TEST_STR_SIZE + OVERLAP_LEN));
+}
+
+
+TEST(string_memmove, left_overlap)
+{
+	uint8_t *const src = mem + OFFSET;
+	uint8_t *dst = mem + (OFFSET - OVERLAP_LEN);
+	/* Move data so that data blocks overlap */
+	ret = memmove(dst, src, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_PTR(dst, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, dst, TEST_STR_SIZE);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR + OVERLAP_LEN, dst + TEST_STR_SIZE, OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, mem, OFFSET - OVERLAP_LEN);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, src + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST(string_memmove, move_nothing)
+{
+	uint8_t *dst = mem + OFFSET;
+	/* Check if any data is moved */
+	ret = memmove(dst, TEST_STR, 0);
+	TEST_ASSERT_EQUAL_PTR(dst, ret);
+	TEST_ASSERT_EQUAL_MEMORY(TEST_STR, dst, TEST_STR_SIZE);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, mem, OFFSET);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, dst + TEST_STR_SIZE, MEM_SIZE - (OFFSET + TEST_STR_SIZE));
+}
+
+
+TEST_SETUP(string_memmove_big)
+{
+	mem_big = (uint8_t *)calloc(MEM_SIZE_BIG, 1);
+	uint8_t *dst = mem_big + OFFSET_BIG;
+	TEST_ASSERT_NOT_NULL(mem_big);
+	ret = memcpy(dst, big_buf, BIG_SIZE);
+	TEST_ASSERT_EQUAL_PTR(dst, ret);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, mem_big, OFFSET_BIG);
+	TEST_ASSERT_EACH_EQUAL_UINT8(0, dst + BIG_SIZE, MEM_SIZE_BIG - (OFFSET_BIG + BIG_SIZE));
+}
+
+
+TEST_TEAR_DOWN(string_memmove_big)
+{
+	free(mem_big);
+}
+
+
+TEST(string_memmove_big, no_overlap)
+{
+	uint8_t *const src = mem_big + OFFSET_BIG;
+	uint8_t *dst = mem_big + OFFSET_BIG + BIG_SIZE + 8;
+	/* Move data so that it don't overlap */
+	int i, j, z, n = BIG_SIZE - 8;
+
+	/* Test all combinations with remainders of division by word for len, dst and src */
+	for (z = 0; z < 8; z++) {
```

Comment:
maybe use meaningful names like `srcOff`, `dstOff`, also introduce local variable inside the loops `int moveSz = n + z` to make the code more readable.

Right now it's hard to verify if the math is correct at a glance.

Also: when doing nested loops usually the outer loop is `i` and the inner loops get `j`, `k`, ... indices - please follow the common schema to avoid potential readability problems.